tsl/src/compression/compression.c

```diff
@@ -1514,3 +1518,164 @@ update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
 		CommandCounterIncrement();
 	}
 }
+
+typedef struct CompressSingleRowState
+{
+	Relation in_rel;
+	Relation out_rel;
+	RowCompressor row_compressor;
+	TupleTableSlot *out_slot;
+} CompressSingleRowState;
+
+static TupleTableSlot *compress_singlerow(CompressSingleRowState *cr, TupleTableSlot *in_slot);
+
+CompressSingleRowState *
+compress_row_init(int srcht_id, Relation in_rel, Relation out_rel)
+{
+	ListCell *lc;
+	List *htcols_list = NIL;
+	int i = 0, cclen;
+	const ColumnCompressionInfo **ccinfo;
+	TupleDesc in_desc = RelationGetDescr(in_rel);
+	TupleDesc out_desc = RelationGetDescr(out_rel);
+	int16 *in_column_offsets;
+	int n_keys;
+	const ColumnCompressionInfo **keys;
+
+	CompressSingleRowState *cr = palloc(sizeof(CompressSingleRowState));
+	cr->out_slot =
+		MakeSingleTupleTableSlotCompat(RelationGetDescr(out_rel), table_slot_callbacks(out_rel));
+	cr->in_rel = in_rel;
+	cr->out_rel = out_rel;
+
+	/* get compression properties for hypertable */
+	htcols_list = ts_hypertable_compression_get(srcht_id);
+	cclen = list_length(htcols_list);
+	ccinfo = palloc(sizeof(ColumnCompressionInfo *) * cclen);
+	foreach (lc, htcols_list)
+	{
+		FormData_hypertable_compression *fd = (FormData_hypertable_compression *) lfirst(lc);
+		ccinfo[i++] = fd;
+	}
+	in_column_offsets =
+		compress_chunk_populate_keys(RelationGetRelid(in_rel), ccinfo, cclen, &n_keys, &keys);
+	row_compressor_init(&cr->row_compressor,
+						in_desc,
+						out_rel,
+						cclen,
+						ccinfo,
+						in_column_offsets,
+						out_desc->natts);
+	return cr;
+}
+
+/* create a single row compressed tuple from data in slot */
+TupleTableSlot *
+compress_row_exec(CompressSingleRowState *cr, TupleTableSlot *slot)
+{
+	TupleTableSlot *compress_slot;
+	slot_getallattrs(slot);
+
+	cr->row_compressor.rows_compressed_into_current_value = 0;
+	row_compressor_update_group(&cr->row_compressor, slot);
+	row_compressor_append_row(&cr->row_compressor, slot);
+	compress_slot = compress_singlerow(cr, slot);
+	return compress_slot;
+}
+
+static TupleTableSlot *
+compress_singlerow(CompressSingleRowState *cr, TupleTableSlot *in_slot)
+{
+	Datum *invalues, *out_values;
+	bool *out_isnull;
+	TupleTableSlot *out_slot = cr->out_slot;
+	RowCompressor *row_compressor = &cr->row_compressor;
+
+	slot_getallattrs(in_slot);
```

Comment:
You have already fetched all attributes in the calling function.

Reply:
fixed.